### DISCUSSION TITLE
BadTimeout

### DISCUSSION THREAD

**Discussion by yasodakrishnav** [eclipse-milo/milo#1244] [General]
"UaException: status=Bad_Timeout, message=timed out waiting for secure channel"

I am getting above exception while reading the values with security mode is Sign and security policy is Basic256. when i tried with none as security mode/policy, i was able to read the value.

Can you help me with this @kevinherron ?

Thanks

**Comment by kevinherron** [comment]
This means the server is not responding to the OpenSecureChannelRequest from your client.

Maybe a Wireshark capture will help, but generally this is something you would have to troubleshoot with the server.

**Reply by yasodakrishnav** [reply]
Thanks kevin.

And i am planning to read each tag value asynchronously using CompletionService, will that be fine or cause any issue to the server.
In this case what will be best value to set RequestTimeout(default is 60 seconds) to client. I have 1000+ tags to read the values

I will be reading value of tags in every 15minutes.

**Reply by kevinherron** [reply]
This is happening before you can even connect, I'm not sure the strategy for reading tags matters at this time.

That said ... is there a reason you are not using subscriptions? Or at least reading the tags in bulk instead of individually?

**Reply by yasodakrishnav** [reply]
above question is not about connecting to the client.

i just want to read data in every 15mins only that's why i dont wanna use subscriptions. and The reason i am not using bulk read is that if any of the tag failed reading it causing all other tags failed to read the values.

**Reply by yasodakrishnav** [reply]
`
private Pair<String, Object> performRead(OpcUaClient client, NodeId nodeId, OpcUaPoint point, DataObject dataObject){
        DataValue dataValue = null;
        try {
            dataValue = client.readValue(0.0, TimestampsToReturn.Both, nodeId).get();
            Object valueObject = OpcUtils.convertToSpecificDataType(dataValue.getValue().getValue(), point.getDataTypeIdentifier());
            return Pair.of(point.getName(), valueObject);
        } catch (InterruptedException | ExecutionException e) {
           dataObject.addFailedPoint(point.getName());
        }
        return null;
    }

    private DataObject getValue(Controller controller, List<Point> fPoints) {
        OpcUaController opcUaController = (OpcUaController) controller;
        DataObject DataObject = new DataObject(opcUaController.getChildJSON(), CONTROLLER_TYPE);
        OpcUaClient client = null;
        try {

            client = OpcUtils.getOpcUaClient(opcUaController);
            LOGGER.info(GET_VALUE + "Client Connected to --> " + opcUaController.getName() + " " + opcUaController.getUrl());

            List<NodeId> nodeIdList = OpcUtils.getNodeIdList(fPoints);
            LOGGER.info(GET_VALUE + "Reading data for controller " + controller.getName());

            // Use a CompletionService to manage asynchronous read requests
            CompletionService<Pair<String, Object>> completionService = new ExecutorCompletionService<>(Executors.newFixedThreadPool(nodeIdList.size()));

            // Submit read requests for each node ID concurrently
            for (int i = 0; i < nodeIdList.size(); i++) {
                NodeId nodeId = nodeIdList.get(i);
                OpcUaPoint opcUaPoint = (OpcUaPoint) fPoints.get(i);
                OpcUaClient finalClient = client;
                completionService.submit(() -> performRead(finalClient, nodeId, opcUaPoint, dataObject));
            }

            for (int i = 0; i < nodeIdList.size(); i++) {
                try {
                    Future<Pair<String, Object>> future = completionService.take();
                    Pair<String, Object> nameVsValue = future.get(Utils.getIntProperty("opcReadValueTimeout", 1), TimeUnit.SECONDS);
                    if(nameVsValue != null){
                        dataObject.addData(nameVsValue.getKey(), nameVsValue.getValue());
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt(); // Reset the interrupted status
                    LOGGER.error(GET_VALUE + "Thread interrupted while waiting for result", e);
                } catch (ExecutionException e) {
                    LOGGER.error(GET_VALUE + "Exception occurred while retrieving value", e);
                }
            }
        } catch (Exception e) {
            LOGGER.error(GET_VALUE + "Exception occurred", e);
            dataObject.setException(e.getMessage());
        } finally {
            if (client != null) {
                try {
                    client.disconnect();
                    LOGGER.info(GET_VALUE + "Client Disconnected to --> " + opcUaController.getName() + " " + opcUaController.getUrl());
                } catch (Exception e) {
                    LOGGER.error(GET_VALUE + "Error disconnecting client", e);
                }
            }
        }

        return dataObject;
    }
    `
    
    this the code that i am currently using to read value one tag at a time.
    
    do you have better solution to read values in bulk?

**Reply by yasodakrishnav** [reply]
will the above solution cause any issue to the server?

**Reply by yasodakrishnav** [reply]
and i am getting below exceptions for other servers, what does it mean and how can i resolve it?
<img width="1116" alt="image" src="https://github.com/eclipse/milo/assets/67819916/f55e008a-cb19-4456-bdc8-1c35d7908685">

**Reply by yasodakrishnav** [reply]
hi @kevinherron can you help me with the above doubts?

**Reply by kevinherron** [reply]
It's not efficient, but if this is the only way you can read all the data from the server then it will have to work.

> and i am getting below exceptions for other servers, what does it mean and how can i resolve it?

It means the server is closing the connection for some reason. I couldn't tell you why. Maybe other servers consider receiving 1000+ concurrent read requests to be a DoS or simply can't handle the load.
